nexus_constructor/geometry/off_geometry.py

```diff
@@ -0,0 +1,233 @@
+        face_sizes = [len(face) for face in self.faces]
+        return [sum(face_sizes[0:i]) for i in range(len(face_sizes))]
+
+    @property
+    def off_geometry(self) -> OFFGeometry:
+        return self
+
+    @property
+    def vertices(self) -> List[QVector3D]:
+        return self._vertices
+
+    @vertices.setter
+    def vertices(self, new_vertices: List[QVector3D]):
+        self._vertices = new_vertices
+
+    @property
+    def faces(self) -> List[List[int]]:
+        return self._faces
+
+    @faces.setter
+    def faces(self, new_faces: List[List[int]]):
+        self._faces = new_faces
+
+
+class OFFGeometryNexus(OFFGeometry):
+    """
+    3D mesh description of the shape of an object, based on the OFF file format.
+    http://download.nexusformat.org/sphinx/classes/base_classes/NXoff_geometry.html
+    """
+
+    def __init__(
+        self,
+        nexus_file: nx.NexusWrapper,
+        group: h5py.Group,
+        units: str = "",
+        file_path: str = "",
+    ):
+        super().__init__(units)
+        self.file = nexus_file
+        self.group = group
+        self._verify_in_file()
+
+        # Source units and file path are retained only for the purpose of populating the edit component window
+        # with the options previously chosen by the user
+        self.units = units
+        self.file_path = file_path
+
+    def _verify_in_file(self):
+        """
+        Check all the datasets and attributes we require are in the NXoff_geometry group
+        """
+        problems = validate_group(
+            self.group,
+            "NXoff_geometry",
+            (
+                ValidateDataset(
+                    "vertices", shape=(None, 3), attributes={"units": None}
+                ),
+                ValidateDataset("winding_order"),
+                ValidateDataset("faces"),
+            ),
+        )
+        if problems:
+            raise NexusFormatError("\n".join(problems))
+
+    @property
+    def winding_order(self):
+        return [point for face in self.faces for point in face]
+
+    @property
+    def winding_order_indices(self):
+        face_sizes = [len(face) for face in self.faces]
+        return [sum(face_sizes[0:i]) for i in range(len(face_sizes))]
+
+    @property
+    def off_geometry(self) -> OFFGeometry:
+        return OFFGeometryNoNexus(self.vertices, self.faces)
+
+    @property
+    def vertices(self) -> List[QVector3D]:
+        vertices_from_file = self.group["vertices"]
+        number_of_vertices = vertices_from_file.shape[0]
+        return [
+            numpy_array_to_qvector3d(vertices_from_file[vertex_number][:])
+            for vertex_number in range(number_of_vertices)
+        ]
+
+    @vertices.setter
+    def vertices(self, new_vertices: List[QVector3D]):
+        record_vertices_in_file(self.file, self.group, new_vertices)
+
+    @property
+    def faces(self) -> List[List[int]]:
+        winding_order_from_file = self.group["winding_order"]
```

Comment:
It's not immediately obvious what this function is doing. Can we have some tests that specifically cover this method?

Reply:
Added `test_can_retrieve_list_of_vertices_for_each_face` to `test_off_geometry.py`